internal/builtinfunctions/functions_test.go

```diff
@@ -819,3 +821,104 @@ func Test_floatToFormattedString_success(t *testing.T) {
+		map[string]any{"loop_id": 3},
+	}
+	repeatedInputs := map[string]any{
+		"a": "A", "b": "B",
+	}
+	repeatedInputName := builtinfunctions.CombinedObjPropertyConstantName
+	itemName := builtinfunctions.CombinedObjPropertyItemName
+	outputExp := []any{
+		map[string]any{itemName: items[0], repeatedInputName: repeatedInputs},
+		map[string]any{itemName: items[1], repeatedInputName: repeatedInputs},
+		map[string]any{itemName: items[2], repeatedInputName: repeatedInputs},
+	}
+	functionToTest := builtinfunctions.GetFunctions()["bindConstants"]
+	output, err := functionToTest.Call([]any{items, repeatedInputs})
+	assert.NoError(t, err)
+	assert.Equals(t, output.([]any), outputExp)
+
+	output, err = functionToTest.Call([]any{[]any{}, repeatedInputs})
+	assert.NoError(t, err)
+	assert.Equals(t, output.([]any), []any{})
+}
+
+func defaultPropertySchema(t schema.Type) *schema.PropertySchema {
+	return schema.NewPropertySchema(t, nil, false, nil, nil, nil, nil, nil)
+}
+
+func joinStrs(s1, s2 string) string {
+	return strings.Join([]string{s1, s2}, builtinfunctions.CombinedObjIDDelimiter)
+}
+
+func TestHandleTypeSchemaZip(t *testing.T) {
+	basicStringSchema := schema.NewStringSchema(nil, nil, nil)
+	basicIntSchema := schema.NewIntSchema(nil, nil, nil)
+	strTypeID := string(basicStringSchema.TypeID())
+	intTypeID := string(basicIntSchema.TypeID())
+	listInt := schema.NewListSchema(basicIntSchema, nil, nil)
+	listStr := schema.NewListSchema(basicStringSchema, nil, nil)
+
+	_, err := builtinfunctions.HandleTypeSchemaCombine(
+		[]schema.Type{basicStringSchema, basicIntSchema})
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "expected first type to be a list schema")
+
+	_, err = builtinfunctions.HandleTypeSchemaCombine(
+		[]schema.Type{listInt})
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "expected exactly two types")
+
+	myFirstObj := schema.NewObjectSchema(
+		"ObjectTitle",
+		map[string]*schema.PropertySchema{
+			"a": defaultPropertySchema(basicStringSchema),
+			"b": defaultPropertySchema(basicStringSchema),
+		})
+	listMyFirstObj := schema.NewListSchema(myFirstObj, nil, nil)
+	constantsObj := schema.NewObjectSchema(
+		"Constants",
+		map[string]*schema.PropertySchema{
+			"p_str": defaultPropertySchema(basicStringSchema),
+			"p_int": defaultPropertySchema(basicIntSchema),
+		})
+
+	type testInput struct {
+		typeArgs       []schema.Type
+		expectedResult string
+	}
+	testInputs := []testInput{
+		{
+			typeArgs:       []schema.Type{listInt, constantsObj},
+			expectedResult: joinStrs(intTypeID, constantsObj.ID()),
+		},
+		{
+			typeArgs:       []schema.Type{listMyFirstObj, constantsObj},
+			expectedResult: joinStrs(myFirstObj.ID(), constantsObj.ID()),
+		},
+		{
+			typeArgs:       []schema.Type{listMyFirstObj, basicStringSchema},
+			expectedResult: joinStrs(myFirstObj.ID(), strTypeID),
+		},
+		{
+			typeArgs:       []schema.Type{listStr, basicIntSchema},
+			expectedResult: joinStrs(strTypeID, intTypeID),
+		},
+	}
```

Comment:
I have no objections to explanatory comments...so long as they add value and don't just parrot what the code says and uselessly take up space.  Obviously, it would be best if the code were expressive enough that comments aren't needed, but based on Jared's reaction, it isn't.  (Sadly, the variable names are supposed to _help_ not get in the way, but I guess that failed....  Since I can't come up with a concrete way to improve the code, let's go with comments.)

A block comment for the whole table might help, but only if it's speaking in reasonably general terms (i.e., listing the individual cases in the header comment is perilous, because it'll probably end up out of sync with the code -- if what we want is an explanation of each case, let's go with putting a brief line comment on each open brace).

Reply:
what would you like to see?

Reply:
test cases added